python/rmm/_lib/memory_resource.pyx

```diff
@@ -271,43 +323,118 @@ cpdef _initialize(
 
-    _set_default_resource(
-        mr
-    )
+    if devices is None:
+        devices = [0]
+    elif isinstance(devices, int):
+        devices = [devices]
 
+    # create a memory resource per specified device
+    for device in devices:
+        set_current_device(device)
+
+        if logging:
+            mr = LoggingResourceAdaptor(typ(*args), log_file_name.encode())
+        else:
+            mr = typ(*args)
 
-cpdef _set_default_resource(MemoryResource mr):
+        _set_per_device_resource(device, mr)
+    
+    # reset CUDA device to original
+    set_current_device(original_device)
+
+
+cpdef get_per_device_resource(int device):
     """
-    Set the memory resource to use for RMM device allocations.
+    Get the default memory resource for the specified device.
 
     Parameters
     ----------
+    device : int
+        The ID of the device for which to get the memory resource.
+    """
+    global _per_device_mrs
+    return _per_device_mrs[device]
+
+
+cpdef _set_per_device_resource(int device, MemoryResource mr):
+    """
+    Set the default memory resource for the specified device.
+
+    Parameters
+    ----------
+    device : int
+        The ID of the device for which to get the memory resource.
     mr : MemoryResource
-        A MemoryResource object. See `rmm.mr` for the different
-        MemoryResource types available.
+        The memory resource to set.
     """
-    global _mr
-    _mr = mr
-    set_default_resource(_mr.c_obj)
+    global _per_device_mrs
+    _per_device_mrs[device] = mr
+    _mr = mr  # coerce Python object to C object 
+    set_per_device_resource(device, _mr.c_obj)
 
 
-cpdef get_default_resource_type():
+cpdef set_current_device_resource(MemoryResource mr):
     """
-    Get the default memory resource type used for RMM device allocations.
+    Set the default memory resource for the current device.
+
+    Parameters
+    ----------
+    mr : MemoryResource
+        The memory resource to set.
     """
-    return type(_mr)
+    _set_per_device_resource(get_current_device(), mr)
+
+
+cpdef get_per_device_resource_type(int device):
+    """
+    Get the memory resource type used for RMM device allocations on the
+    specified device.
+
+    Parameters
+    ----------
+    device : int
+        The device ID
+    """
+    return type(get_per_device_resource(device))
+
+
+cpdef get_current_device_resource():
+    return get_per_device_resource(get_current_device())
+
+
+cpdef get_current_device_resource_type():
+    """
+    Get the memory resource type used for RMM device allocations on the
+    current device.
+    """
+    return type(get_current_device_resource())
 
 
 cpdef is_initialized():
-    global _mr
-    return _mr.c_obj.get() is not NULL
+    """
+    Check whether RMM is initialized
+    """
+    global _per_device_mrs
+    cdef MemoryResource each_mr
+    return all(
+        [each_mr.c_obj.get() is not NULL
+            for each_mr in _per_device_mrs.values()]
```

Comment:
I think we can remove the brackets here as this just unnecessarily creates an intermediate Python list instead of using a generator expression:
```suggestion
        each_mr.c_obj.get() is not NULL
        for each_mr in _per_device_mrs.values()
```

Reply:
`closures inside cpdef functions not yet supported`

Reply:
`Invalid types for 'is_not' (Python object, void *)`

Reply:
well then, looks like we stay with the Python list for now 😅